pkg/kubelet/pleg/evented.go

```diff
@@ -173,22 +174,23 @@ func (e *EventedPLEG) Healthy() (bool, error) {
 	return true, nil
 }
 
-func (e *EventedPLEG) watchEventsChannel() {
+func (e *EventedPLEG) watchEventsChannel(ctx context.Context) {
 	containerEventsResponseCh := make(chan *runtimeapi.ContainerEventResponse, cap(e.eventChannel))
 	defer close(containerEventsResponseCh)
 
+	logger := klog.FromContext(ctx)
 	// Get the container events from the runtime.
 	go func() {
 		numAttempts := 0
 		for {
 			if numAttempts >= e.eventedPlegMaxStreamRetries {
 				if isEventedPLEGInUse() {
 					// Fall back to Generic PLEG relisting since Evented PLEG is not working.
-					klog.V(4).InfoS("Fall back to Generic PLEG relisting since Evented PLEG is not working")
+					logger.V(4).Info("Fall back to Generic PLEG relisting since Evented PLEG is not working")
 					e.Stop()
 					e.genericPleg.Stop()       // Stop the existing Generic PLEG which runs with longer relisting period when Evented PLEG is in use.
 					e.Update(e.relistDuration) // Update the relisting period to the default value for the Generic PLEG.
-					e.genericPleg.Start()
+					e.genericPleg.Start(ctx)
```

Comment:
/lgtm cancel

There's a subtle bug here: this ctx is the one for `watchEventsChannel`, which in turn is the one which was tied to the one created by `context.WithCancelCause(ctx)` in `Start`. In other words, it is the one which got canceled when calling `e.Stop()`, so this `Start` here will not actually run anything.

`Start` must store it's `ctx` parameter in a `startCtx` field and use it here when it (re)starts `e.genericPleg`.

@harche, @pacoxu: is there any test for this particular code area?

Reply:
Is enabling context support (i.e. using `context.Context`) worth it?

The simpler alternative is to provide a `klog.Logger` for the instance and use that everywhere.

Reply:
```
/test pull-kubernetes-e2e-kind-beta-features
/test pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e
/test pull-kubernetes-e2e-kind-evented-pleg
```
for the evented pleg, other tests should be always using the generic PLEG.

Reply:
> Is enabling context support (i.e. using `context.Context`) worth it?

I think it worth it. It makes the code clear and consistent in my opinion. Passing logger around to avoid fixing 2 stop channels mess we've discovered here would be a last resort. I'd suggest to fix the context issue as you've proposed above.